rfc/rfc-0000-common-blueprint.md

```diff
@@ -0,0 +1,220 @@
+      inputs:
+        - resource: zero-resource
+          output: zoo
+          name: zoo-resource-name
+      inline:
+        # mutually exclusive - template, ytt
+        template: {}
+        ytt: {}
+
+        outputs:
+          - name: foo
+            valuePath: .status.url
+
+    - name: second-resource
+      # mutually exclusive - blueprintName, options
+      options:
+        - name: another-blueprint
+        - name: diff-blueprint
+      inputs:
+        - resource: first-resource
+          output: foo
+          name: a-new-resource-name
+        - resource: zero-resource
+          output: zoo
+          name: a-new-resource-name
+
+    - name: third-resource
+      # mutually exclusive - blueprintName, options
+      blueprintName: another-another-blueprint
+      inputs:
+        - resource: first-resource
+          output: foo
+          name: a-new-resource-name
+        - resource: second-resource
+          output: bar
+          name: another-new-name
+  
+  outputs:
+    - name: baz
+      valuePath: resources[first-resource].outputs[foo]
+    - name: fizz
+      valuePath: resources[second-resource].outputs[bar]
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterBlueprintType
+metadata:
+  name: my-baz-fizz-type
+spec:
+  outputs:
+    - name: baz
+    - name: fizz
+  
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterBlueprintSelector
+metadata:
+  name: my-selector
+spec:
+  blueprintRef:
+    name: my-blueprint
+  ownerType:
+    apiVersion: carto.run/v1alpha1
+    kind: Workload
+  selector:
+    matchLabels: []
+    matchFields: []
+    matchExpressions: []
+```
+
+Example with template, one output, and deliverable selector:
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: ClusterBlueprint
+metadata:
+  name: my-blueprint-2
+spec:
+  blueprintTypeRef:
+    name: my-url-type
+
+  # mutually exclusive - resources, template, ytt (i.e. a blueprint is either a template or a supply chain)
+  template:
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: test
+    data:
+      foo: bar
+
+  outputs:
+    - name: url
+      valuePath: .data.foo
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterBlueprintType
+metadata:
+  name: my-url-type
+spec:
+  outputs:
+    - name: url
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterBlueprintSelector
+metadata:
+  name: my-selector-2
+spec:
+  blueprintRef:
+    name: my-blueprint-2
+  ownerType:
+    apiVersion: carto.run/v1alpha1
+    kind: Deliverable
+  selector:
+    matchLabels: []
+    matchFields: []
+    matchExpressions: []
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+This RFC does not propose any new functionality (other than snippets), but rather proposes re-defining the API and CR's to encapsulate the same functionality that exists today.
+
+
+# Migration
+[migration]: #migration
+
+This RFC proposes all new CRDs, so there is no need for migration.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
```

Comment:
You could introduce these new CRDs as an experimental extension to handle cases that the existing resources can't handle. This would give you hands-on experience and the ability to change their API until they're a suitable replacement for the existing resources. At that point you could do a major version bump and remove the original resources